datumaro/components/media.py

```diff
@@ -178,3 +179,255 @@ def save(self, path):
+                yield VideoFrame(video=self, index=index)
+        else:
+            # Need to decode to iterate over frames
+            yield from self._get_iterator()
+
+    @property
+    def frame_count(self) -> Optional[int]:
+        """
+        Returns frame count, if video provides such information.
+        Note that not all videos provide length / duration metainfo.
+
+        The count is affected by the frame filtering options of the object.
+        """
+        return self._get_length()
+
+    @property
+    def frame_size(self) -> Tuple[int, int]:
+        if self._frame_size is None:
+            self._frame_size = self._get_frame_size()
+        return self._frame_size
+
+    def _get_frame_size(self) -> Tuple[int, int]:
+        w = self._container.get(cv2.CAP_PROP_FRAME_WIDTH)
+        h = self._container.get(cv2.CAP_PROP_FRAME_HEIGHT)
+
+        if h and w:
+            frame_size = (int(h), int(w))
+        else:
+            image = next(self._get_iterator()).data
+            frame_size = image.shape[0:2]
+
+        return frame_size
+
+    @staticmethod
+    def _get_frame_count(container) -> Optional[int]:
+        # Not all videos provide length / duration metainfo
+        # Note that this information can be invalid or inaccurate
+        # due to variable frame rate
+        # https://stackoverflow.com/a/47796468
+        video_length = container.get(cv2.CAP_PROP_FRAME_COUNT)
+        if video_length:
+            video_length = int(video_length)
+        else:
+            video_length = None
+        return video_length
+
+    def _get_end_frame(self):
+        if self._end_frame is not None and self._frame_count is not None:
+            end_frame = min(self._end_frame, self._frame_count)
+        else:
+            end_frame = self._end_frame or self._frame_count
+
+        return end_frame
+
+    def _get_length(self):
+        if self._length is None:
+            end_frame = self._get_end_frame()
+
+            length = None
+            if end_frame is not None:
+                length = (end_frame - self._start_frame) // self._step
+                assert 0 < length
+
+            self._length = length
+
+        return self._length
+
+    def _includes_frame(self, i):
+        end_frame = self._get_end_frame()
+        if self._start_frame <= i:
+            if (i - self._start_frame) % self._step == 0:
+                if end_frame is None or i < end_frame:
+                    return True
+
+        return False
+
+    def _get_iterator(self):
+        if self._iterator is None:
+            self._iterator = _VideoFrameIterator(self)
+        return self._iterator
+
+    def _get_container(self):
+        if self._container is None:
+            self._reset()
+        return self._container
+
+    def _open(self):
+        return cv2.VideoCapture(self._path)
+
+    def _reset(self):
+        if self._container is not None:
+            self._container.release()
+        self._container = self._open()
```

Comment:
It was a term from pyav https://pyav.org/docs/develop/api/container.html. I'm ok with `_readed`, renamed. I don't want to depend on OpenCV naming here, because it is not very explanatory in this context.